### ISSUE TITLE
Candidate on CI: implement `latest` or commit slug in the link

### ISSUE THREAD

**Issue by mgajda** [haskell/hackage-server#861]
I have made a nice release process for Hackage packages using GitLab CI _(example run: https://gitlab.com/migamake/homplexity/pipelines/120317820, source code: https://gitlab.com/migamake/homplexity/-/blob/master/.gitlab-ci.yml)_

I even upload a candidate to make sure it is release ready.
However, I cannot put the link to the current candidate, since CI can only use Git Commit hash or branch name as variable name.
So I need to go into candidate upload job log, like here: https://gitlab.com/migamake/homplexity/-/jobs/447092421
And find the link myself, or try to recall which version of the package is produced:
https://hackage.haskell.org/package/homplexity-0.4.6.0/candidate

Since some CDs allow for URL to be included, it would be fixed by either:
1. Providing https://hackage.haskell.org/package/homplexity-latest/candidate link
2. Or allowing one to put upload hash for the candidate 
```
cabal upload homplexity.tar.gz --candidate-hash c90c0148
```
https://hackage.haskell.org/package/homplexity-c90c0148/candidate

There is no necessity to store it longer than few weeks, or when new candidate is uploaded, but it would allow one to double check package description, and documentation before release.

That would be related to #597, and possibly allow to implement #558 better with the help of external CI/CD.

**Comment by phadej** [comment]
Let me ask a question to be sure I understand right:

The GitLab CI would be able to provide a link in UI to the uploaded artifacts, but the URL have to be constructible from *external* information, i.e. either hard-coded info (e.g. envvar `PACKAGENAME=foo`), or git commit/branch  (`c90c0148` or `master`) or GitLab CI job identifiers (say `447092421`); but there's **no** way that can depend on the contents of the repository (e.g. if there were a way to extract package version from the "inside")?

And a follow-up question: how it would look like in a pipeline summary like https://gitlab.com/migamake/homplexity/pipelines/120317820, I don't see anything I could click (except the final Hackage deploy job.

**Comment by mgajda** [comment]
Yes. The CI/CD variables are either global, defined in CI script, or given by a trigger that restarts a build process, so indeed there is no way to depend on the contents on the repo there.
(It is even more involved with system like Shippable, that separates CI configuration from the repo.)

Just for you, here is an demo that uses hard-wired version number at the moment:
https://gitlab.com/migamake/homplexity/-/jobs/448777451
(Documented here: https://docs.gitlab.com/ee/ci/environments.html#using-the-environment-url)
But I'll make a ticket to show it directly in the pipeline view.

_Test environments are present on other CIs as well. Similar feature on Shippable:_
http://blog.shippable.com/how-to-create-on-demand-test-environments

**Comment by phadej** [comment]
https://gitlab.com/migamake/homplexity/-/environments returns 404, probably some access control; do you mind just posting a screenshot?

**Comment by mgajda** [comment]
[This is the `/-/environments`](https://imgur.com/a/cNS90tW)

**Comment by mgajda** [comment]
Links to [GitLab feature request](https://gitlab.com/gitlab-org/gitlab/issues/207963).
